### ISSUE TITLE
Templated jinja is not ignored

### ISSUE THREAD

**Issue by mdimercurio** [sqlfluff/sqlfluff#831]
When trying to use macros from a package, sqlfluff returns an error when these should be skipped because of the `ignore_templated_areas` parameter.

## Expected Behaviour

`dbt_utils` macros should be ignored in the lint command.

## Observed Behaviour

`sqlfluff lint` output

```
L:   - | P:   - |  TMP | Unrecoverable failure in Jinja templating: 'dbt_utils'
                       | is undefined. Have you configured your variables?
                       | https://docs.sqlfluff.com/en/latest/configuration.html
L:   4 | P:   4 |  TMP | Undefined jinja template variable: 'dbt_utils'
```
-- If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant. Finally, please add any labels that are related to the issue before submitting so that it can be addressed easily! --

## Steps to Reproduce

Query:
```sql
select id
    , count(*)
from {{ source("snowplow", "events") }}
{{ dbt_utils.group_by(1) }}
```

Try to lint using:
> sqlfluff lint models/snowplow_events.sql
output:
```
== [models/snowplow_events.sql] FAIL
L:   - | P:   - |  TMP | Unrecoverable failure in Jinja templating: 'dbt_utils'
                       | is undefined. Have you configured your variables?
                       | https://docs.sqlfluff.com/en/latest/configuration.html
L:   4 | P:   4 |  TMP | Undefined jinja template variable: 'dbt_utils'
```

## Version
Include the output of `sqlfluff --version` along with your Python version

python 3.8.3

```
> sqlfluff --version
sqlfluff, version 0.4.1
```

## Configuration
```
[sqlfluff]
templater = jinja
ignore_templated_areas = True
```

**Comment by NiallRees** [comment]
> When trying to use macros from a package, sqlfluff returns an error when these should be skipped because of the `ignore_templated_areas` parameter.
> 
> ## Expected Behaviour
> `dbt_utils` macros should be ignored in the lint command.
> 
> ## Observed Behaviour
> `sqlfluff lint` output
> 
> ```
> L:   - | P:   - |  TMP | Unrecoverable failure in Jinja templating: 'dbt_utils'
>                        | is undefined. Have you configured your variables?
>                        | https://docs.sqlfluff.com/en/latest/configuration.html
> L:   4 | P:   4 |  TMP | Undefined jinja template variable: 'dbt_utils'
> ```
> 
> -- If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant. Finally, please add any labels that are related to the issue before submitting so that it can be addressed easily! --
> 
> ## Steps to Reproduce
> Query:
> 
> ```sql
> select id
>     , count(*)
> from {{ source("snowplow", "events") }}
> {{ dbt_utils.group_by(1) }}
> ```
> 
> Try to lint using:
> 
> ```shell
> sqlfluff lint models/snowplow_events.sql
> ```
> 
> output:
> 
> ```
> == [models/snowplow_events.sql] FAIL
> L:   - | P:   - |  TMP | Unrecoverable failure in Jinja templating: 'dbt_utils'
>                        | is undefined. Have you configured your variables?
>                        | https://docs.sqlfluff.com/en/latest/configuration.html
> L:   4 | P:   4 |  TMP | Undefined jinja template variable: 'dbt_utils'
> ```
> 
> ## Version
> Include the output of `sqlfluff --version` along with your Python version
> 
> python 3.8.3
> 
> ```
> > sqlfluff --version
> sqlfluff, version 0.4.1
> ```
> 
> ## Configuration
> ```
> [sqlfluff]
> templater = jinja
> ignore_templated_areas = True
> ```

Hey @mdimercurio! So this is actually the expected behaviour. ignore_templated_areas just means that linting errors directly attributable to a templates section are not raised. The best way to overcome the error you're seeing is with the dbt templater 🙂 I'm going to close this issue if ok 👍

**Comment by mdimercurio** [comment]
Hi @NiallRees, I opened this issue after discussing it on [Slack here](https://sqlfluff.slack.com/archives/C01GX4L213J/p1614473146020900). When using the dbt templater, I'm getting a lot of failures due to jinja adding line breaks and trailing spaces. Is there a solution to resolve that if I should use the dbt templater?

**Comment by pwildenhain** [comment]
The `dbt` templater should ignore the lint errors like Nial said, so this would still be a relevant issue. Can you post what the output looks like when you set `templater = dbt`?

**Comment by mdimercurio** [comment]
here is an example of my issue with the templater:
```sql
{%
    set properties = {
        "id": "id",
        "type": "post_type",
        "channel_id": "episode_id"
    }
%}

select
    {% for prop, col in properties.items() %}
        {% if not loop.first %} , {% endif %}
            {{prop}} as {{ col}}
    {% endfor %}
from {{ ref("snowplow_events_dev") }}
```

```
== [models/snowplow_bm_webhook_post_received.sql] FAIL
L:   8 | P:  10 | L003 | Indentation not consistent with line #6
L:   8 | P:  10 | L019 | Found trailing comma. Expected only leading.
L:  11 | P:  10 | L003 | Indentation not hanging or a multiple of 4 spaces
L:  11 | P:  10 | L019 | Found trailing comma. Expected only leading.
L:  13 | P:   5 | L003 | Indentation not consistent with line #3
L:  15 | P:  20 | L009 | Files must end with a trailing newline.
```

**Comment by alanmcruickshank** [comment]
@mdimercurio I've been digging into this issue. I've fixed some elements of it in #892 , namely that the line numbers here are still referring to the file _post templating_ rather than _pre templating_, and that we're not linting the jinja tags.

Now if I lint the same query:
```sql
{%
    set properties = {
        "id": "id",
        "type": "post_type",
        "channel_id": "episode_id"
    }
%}

select
    {% for prop, col in properties.items() %}
        {% if not loop.first %} , {% endif %}
            {{prop}} as {{ col}}
    {% endfor %}
from {{ ref("snowplow_events_dev") }}
```
I get:
```bash
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:  11 | P:  33 | L003 | Indentation not consistent with line #6
L:  11 | P:  33 | L003 | Indentation not hanging or a multiple of 4 spaces
L:  11 | P:  33 | L019 | Found leading comma. Expected only trailing.
L:  11 | P:  33 | L019 | Found leading comma. Expected only trailing.
L:  12 | P:  13 | L046 | Jinja tags should have a single whitespace on either
                       | side.
L:  12 | P:  25 | L046 | Jinja tags should have a single whitespace on either
                       | side.
```

- The L046 ones are new and are due to the whitespace around the `{{prop}}` and `{{ col}}` tags.
- The comma errors are to be expected, notably because the commas themselves aren't generated by macros, they're literals within a loop and so currently not ignored by the `ignore_templated_areas` config.
- Indentation is a bit odd within the loop of your query as I think the `{{prop}}` and `{% if not ....` sections at the same level logically. 

If I correct the query to this (and update my config to use leading commas) I get this query:

`.sqlfluff`
```ini
[sqlfluff:rules]
comma_style = leading
```

`test.sql`
```sql
{%
    set properties = {
        "id": "id",
        "type": "post_type",
        "channel_id": "episode_id"
    }
%}

select
    {% for prop, col in properties.items() %}
        {% if not loop.first %}, {% endif %}{{ prop }} as {{ col }}
    {% endfor %}
from {{ ref("snowplow_events_dev") }}
```

Then I get no linting errors. *However* I realize this might not be optimal for readability because of having the comma code on the same line. An alternative would be:

```sql
{%
    set properties = {
        "id": "id",
        "type": "post_type",
        "channel_id": "episode_id"
    }
%}

select
    {% for prop, col in properties.items() %}
        {% if not loop.first %}, {% endif %}  -- noqa:L019,L003
        {{ prop }} as {{ col }}
    {% endfor %}
from {{ ref("snowplow_events_dev") }}
```

Which also gets no linting errors.

--

Going back to your original question. I think we probably need to be clearer on what `ignore_templated_areas` actually does. The current functionality ignores any errors found _within sections of code coming directly from jinja tags` (e.g. from within `{{ prop }}`), but still keeps errors from literal code found _within loops_, like the example you've posted.

I think this is then working as expected but it poses two questions:
1. Should we have an alternate configuration where we can *also* ignore any errors within looped sections?
2. Does the documentation need to be clearer around `ignore_templated_areas` so that it's clearer what it actually does?
